{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>- for testing in different (commercial) environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "G4yBrceuFbf3",
    "tags": []
   },
   "source": [
    "<img src=\"https://raw.githubusercontent.com/sokrypton/ColabFold/main/.github/ColabFold_Marv_Logo_Small.png\" height=\"200\" align=\"right\" style=\"height:240px\">\n",
    "\n",
    "\n",
    "# General instructions and introduction\n",
    "Welcome to the AlphaFold tutorial where we will use a modified version of ColabFold to utilize AlphaFold2 using MMseqs2. Here, you can predict the protein folding structure solely based on the amino acid sequence.\n",
    "- ColabFold is an emerging protein folding prediction tool based on Google DeepMind‚Äôs AlphaFold. \n",
    "- Throughout the workshop there will be questions to be answered, so you get a deeper understanding of what you are calculating\n",
    "- You can run each cell (box) by pressing `shift+enter` while being inside of the cell or by pressing on the triangled play botton above.\n",
    "\n",
    "## ColabFold: AlphaFold2 using MMseqs2\n",
    "Easy to use protein structure and complex prediction using [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) and [Alphafold2-multimer](https://www.biorxiv.org/content/10.1101/2021.10.04.463034v1). Sequence alignments/templates are generated through [MMseqs2](mmseqs.com) and [HHsearch](https://github.com/soedinglab/hh-suite). For more details about the original Jupyter notebook and ColabFold, see <a href=\"#Instructions\">bottom</a> of the notebook, check the [ColabFold GitHub](https://github.com/sokrypton/ColabFold) and read the manuscript. Old versions: [v1.0](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.0-alpha/AlphaFold2.ipynb), [v1.1](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.1-premultimer/AlphaFold2.ipynb), [v1.2](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.2.0/AlphaFold2.ipynb), [v1.3](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.3.0/AlphaFold2.ipynb)\n",
    "\n",
    "[Mirdita M, Sch√ºtze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M. ColabFold: Making protein folding accessible to all.\n",
    "*Nature Methods*, 2022](https://www.nature.com/articles/s41592-022-01488-1) \n",
    "\n",
    "\n",
    "## Learning objectives\n",
    "\n",
    "<img src=\"https://media.nature.com/w1248/magazine-assets/d41586-022-00997-5/d41586-022-00997-5_20300292.gif?as=webp\" height=\"200\" align=\"right\" style=\"height:240px\">\n",
    "\n",
    "After completing this workshop you will be able to:\n",
    " - Describe how to use ColabFold at UCloud to predict protein structures.\n",
    " - Explain and interpret the results generated using ColabFold.\n",
    " - Use ColabFold to predict the protein structure of any specific protein of interest.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "\n",
    "(Double-click to edit the cell) <br>\n",
    "‚ùî __Question I: <u> What is AlphaFold and why does it play a crucial role in structural biology?</u>__ <br>\n",
    "_Answer:_  \n",
    "<br>\n",
    "‚ùî __Question II: <u> What is ColabFold and how does it differ from AlphaFold2?</u>__ <br>\n",
    "_Answer:_  \n",
    "<br>\n",
    "‚ùî __Question III: <u> What is MMseqs2 and how does it work?</u>__ <br>\n",
    "_Answer:_  \n",
    "<br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "üí° Useful link(s): <br>\n",
    "[Highly accurate protein structure prediction with AlphaFold](https://www.nature.com/articles/s41586-021-03819-2) <br>\n",
    "[How AlphaFold can realize AI‚Äôs full potential in structural biology](https://www.nature.com/articles/d41586-022-02088-x) <br>\n",
    "[ColabFold: making protein folding accessible to all](https://www.nature.com/articles/s41592-022-01488-1) <br>\n",
    "[MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets](https://www.nature.com/articles/nbt.3988)<br>\n",
    "[ColabFold at GitHub](https://github.com/sokrypton/ColabFold)<br>\n",
    "[YouTube: ColabFold - Making protein folding accessible to all via Google Colab!](https://youtu.be/Rfw7thgGTwI)<br>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization\n",
    "Please run the following cell to initialize ColabFold by defining the paths of the working environment to be used for running ColabFold at UCloud. __This is needed only once__. üöÄ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys,os\n",
    "\n",
    "CONDA_DIR=\"/work/SW/CondaEnvironments/colabfold/\"\n",
    "CONDA_BIN=os.path.join(CONDA_DIR, \"bin\")\n",
    "CONDA_LIB=os.path.join(CONDA_DIR, \"lib64\")\n",
    "\n",
    "if CONDA_BIN not in os.environ[\"PATH\"]:\n",
    "    os.environ[\"PATH\"] = ':'.join((CONDA_BIN, os.environ[\"PATH\"]))\n",
    "\n",
    "os.environ[\"LD_LIBRARY_PATH\"] = \"\"\n",
    "if CONDA_LIB not in os.environ[\"LD_LIBRARY_PATH\"]:\n",
    "    os.environ[\"LD_LIBRARY_PATH\"] = ':'.join((CONDA_LIB, os.environ[\"LD_LIBRARY_PATH\"]))\n",
    "sys.path.insert(1, '/work/SW/ColabFold/ColabFold_software')\n",
    "\n",
    "print(sys.version)\n",
    "print(\"PATH=\" + os.environ[\"PATH\"])\n",
    "print(\"LD_LIBRARY_PATH=\" + os.environ[\"LD_LIBRARY_PATH\"])\n",
    "print(\"sys.path=\" + str(sys.path))\n",
    "print(\"PWD=\" + os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "<img src=\"https://upload.wikimedia.org/wikipedia/en/e/e5/UniProt_%28logo%29.png\" height=\"200\" align=\"right\" style=\"height:100px\">\n",
    "\n",
    "## Selecting an example protein\n",
    "Now you can select an amino acid sequence of interest to be used as the query sequence. </br>\n",
    "In this workshop, we will use a sequence from the protein [PPARŒ≥ (peroxisome proliferator activated receptor Œ≥)](https://www.uniprot.org/uniprotkb/P37231/entry) obtained from the protein database [UniProt](https://www.uniprot.org/). <br>\n",
    "UniProt is the world's leading high-quality, comprehensive and freely accessible resource of protein sequence and functional information. <br>\n",
    "Here, you can find functional information on any annotated protein of interest:\n",
    "1. Go to [uniprot.org](https://www.uniprot.org/).\n",
    "2. Search for your favorite protein in the search bar.\n",
    "3. Click on the entry ID in the column to the left.\n",
    "4. You can find different functional informations about the protein in the menu to the left.\n",
    "5. Click on __Sequence & Isoforms__ in the menu.\n",
    "6. Click on __Copy sequence__ in the grey box containing the amino acid sequence to copy the entire sequence without any gaps.\n",
    "7. You can now insert this sequence as query sequence for predict the protein structure using ColabFold.\n",
    "\n",
    "üí° Useful link(s): <br>\n",
    "[UniProt: the universal protein knowledgebase in 2021](https://academic.oup.com/nar/article/49/D1/D480/6006196?login=true) <br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "\n",
    "(Double-click to edit the cell) <br>\n",
    "‚ùî __Question IV: <u> What is your favorite protein and what is the amino acid sequence?</u>__ <br>\n",
    "_Answer:_  \n",
    "_Sequence:_  \n",
    "<br>\n",
    "‚ùî __Question V: <u> What is the sequence for PPARŒ≥?</u>__ <br>\n",
    "_Sequence:_  \n",
    "<br>\n",
    "‚ùî __Question VI: <u> What is the function of PPARŒ≥ and which biological processes is the protein involved in?</u>__ <br>\n",
    "_Answer:_  \n",
    "<br>\n",
    "‚ùî __Question VII: <u> Name 3 different mutations in the amino acid sequence of PPARŒ≥ and the disease(s) they can lead to.</u>__ <br>\n",
    "_Answer:_  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "üí° Useful link(s): <br>\n",
    "[UniProt](https://www.uniprot.org/) <br>\n",
    "[PPARgamma in Metabolism, Immunity, and Cancer: Unified and Diverse Mechanisms of Action](https://pubmed.ncbi.nlm.nih.gov/33716977/) <br>\n",
    "[Review of the Structural and Dynamic Mechanisms of PPARŒ≥ Partial Agonism](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4578752/) <br>\n",
    "[Molecular Mechanisms and Genome-Wide Aspects of PPAR Subtype Specific Transactivation](https://pubmed.ncbi.nlm.nih.gov/20862367/) <br>\n",
    "<br>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameter settings\n",
    "In the following cell, you find the parameters used to predict the structure. They can all be set and adjusted. <br>\n",
    "For this workshop we will use the default parameters meaning that you can run the next cell without any further adjustments.\n",
    "\n",
    "Or you change the parameters and rerun the cell. We recommend trying out another protein sequence.\n",
    "\n",
    "_Note: Longer sequences lead to longer run times!_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "############ Collected parameter values\n",
    "\n",
    "### Biological parameters\n",
    "## Use `:` to specify inter-protein chainbreaks for **modeling complexes** (supports homo- and hetero-oligomers). For example **PI...SK:PI...SK** for a homodimer\n",
    "query_sequence = 'MGETLGDSPIDPESDSFTDTLSANISQEMTMVDTEMPFWPTNFGISSVDLSVMEDHSHSFDIKPFTTVDFSSISTPHYEDIPFTRTDPVVADYKYDLKLQEYQSAIKVEPASPPYYSEKTQLYNKPHEEPSNSLMAIECRVCGDKASGFHYGVHACEGCKGFFRRTIRLKLIYDRCDLNCRIHKKSRNKCQYCRFQKCLAVGMSHNAIRFGRMPQAEKEKLLAEISSDIDQLNPESADLRALAKHLYDSYIKSFPLTKAKARAILTGKTTDKSPFVIYDMNSLMMGEDKIKFKHITPLQEQSKEVAIRIFQGCQFRSVEAVQEITEYAKSIPGFVNLDLNDQVTLLKYGVHEIIYTMLASLMNKDGVLISEGQGFMTREFLKSLRKPFGDFMEPKFEFAVKFNALELDDSDLAIFIAVIILSGDRPGLLNVKPIEDIQDNLLQALELQLKLNHPESSQLFAKLLQKMTDLRQIVTEHVQLLQVIKKTETDMSLHPLLQEIYKDLY' #@param {type:\"string\"}\n",
    "\n",
    "### Technical parameters\n",
    "\n",
    "## MSA options (custom MSA upload, single sequence, pairing mode)\n",
    "# Different values: \"MMseqs2 (UniRef+Environmental)\", \"MMseqs2 (UniRef only)\",\"single_sequence\",\"custom\"\n",
    "msa_mode = \"MMseqs2 (UniRef+Environmental)\" \n",
    "# Different values: \"unpaired+paired\",\"paired\",\"unpaired\"\n",
    "# \"unpaired+paired\" = pair sequences from same species + unpaired MSA, \"unpaired\" = seperate MSA for each chain, \"paired\" - only use paired sequences.\n",
    "pair_mode = \"unpaired+paired\"\n",
    "\n",
    "## number of models to use\n",
    "# Either True or False\n",
    "use_amber = False \n",
    "# Different values: \"none\", \"pdb70\",\"custom\"\n",
    "# \"none\" = no template information is used, \"pdb70\" = detect templates in pdb70, \"custom\" - upload and search own templates (PDB or mmCIF format, see [notes below](#custom_templates))\n",
    "template_mode = \"none\"\n",
    "\n",
    "# Different values: \"auto\", \"AlphaFold2-ptm\", \"AlphaFold2-multimer-v1\", \"AlphaFold2-multimer-v2\"\n",
    "# \"auto\" = protein structure prediction using \"AlphaFold2-ptm\" and complex prediction \"AlphaFold-multimer-v2\". For complexes \"AlphaFold-multimer-v[1,2]\" and \"AlphaFold-ptm\" can be used.model_type = \"auto\" \n",
    "\n",
    "model_type = \"auto\"\n",
    "num_recycles = 3\n",
    "\n",
    "\n",
    "### Visualization\n",
    "# Usually a number between 1 and 5\n",
    "rank_num = 1 \n",
    "# Different values: \"chain\", \"lDDT\", \"rainbow\"\n",
    "color = \"lDDT\" \n",
    "# Either True of False\n",
    "show_sidechains = False\n",
    "show_mainchains = False\n",
    "\n",
    "\n",
    "## set dpi for image resolution\n",
    "dpi = 200 \n",
    "\n",
    "########## End parameter settings\n",
    "\n",
    "import os.path\n",
    "import re\n",
    "import hashlib\n",
    "import random\n",
    "\n",
    "def add_hash(x,y):\n",
    "  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n",
    "\n",
    "\n",
    "# remove whitespaces\n",
    "query_sequence = \"\".join(query_sequence.split())\n",
    "\n",
    "jobname = 'test' #@param {type:\"string\"}\n",
    "# remove whitespaces\n",
    "basejobname = \"\".join(jobname.split())\n",
    "basejobname = re.sub(r'\\W+', '', basejobname)\n",
    "jobname = add_hash(basejobname, query_sequence)\n",
    "while os.path.isfile(f\"{jobname}.csv\"):\n",
    "  jobname = add_hash(basejobname, ''.join(random.sample(query_sequence,len(query_sequence))))\n",
    "\n",
    "with open(f\"{jobname}.csv\", \"w\") as text_file:\n",
    "    text_file.write(f\"id,sequence\\n{jobname},{query_sequence}\")\n",
    "\n",
    "queries_path=f\"{jobname}.csv\"\n",
    "\n",
    "if template_mode == \"pdb70\":\n",
    "  use_templates = True\n",
    "  custom_template_path = None\n",
    "elif template_mode == \"custom\":\n",
    "  custom_template_path = f\"{jobname}_template\"\n",
    "  os.mkdir(custom_template_path)\n",
    "  uploaded = files.upload()\n",
    "  use_templates = True\n",
    "  for fn in uploaded.keys():\n",
    "    os.rename(fn, f\"{jobname}_template/{fn}\")\n",
    "else:\n",
    "  custom_template_path = None\n",
    "  use_templates = False\n",
    "\n",
    "    \n",
    "# decide which a3m to use\n",
    "if msa_mode.startswith(\"MMseqs2\"):\n",
    "  a3m_file = f\"{jobname}.a3m\"\n",
    "elif msa_mode == \"custom\":\n",
    "  a3m_file = f\"{jobname}.custom.a3m\"\n",
    "  if not os.path.isfile(a3m_file):\n",
    "    custom_msa_dict = files.upload()\n",
    "    custom_msa = list(custom_msa_dict.keys())[0]\n",
    "    header = 0\n",
    "    import fileinput\n",
    "    for line in fileinput.FileInput(custom_msa,inplace=1):\n",
    "      if line.startswith(\">\"):\n",
    "         header = header + 1\n",
    "      if not line.rstrip():\n",
    "        continue\n",
    "      if line.startswith(\">\") == False and header == 1:\n",
    "         query_sequence = line.rstrip()\n",
    "      print(line, end='')\n",
    "\n",
    "    os.rename(custom_msa, a3m_file)\n",
    "    queries_path=a3m_file\n",
    "    print(f\"moving {custom_msa} to {a3m_file}\")\n",
    "else:\n",
    "  a3m_file = f\"{jobname}.single_sequence.a3m\"\n",
    "  with open(a3m_file, \"w\") as text_file:\n",
    "    text_file.write(\">1\\n%s\" % query_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run protein structure prediction\n",
    "Now you are ready to run the prediction of the protein in the cell below. This can take a while. ‚åö"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "_sztQyz29DIC",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#@title Run Prediction\n",
    "import sys\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "from colabfold.download import download_alphafold_params, default_data_dir\n",
    "from colabfold.utils import setup_logging\n",
    "from colabfold.batch import get_queries, run, set_model_type\n",
    "K80_chk = !nvidia-smi | grep \"Tesla K80\" | wc -l\n",
    "if \"1\" in K80_chk:\n",
    "  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n",
    "  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n",
    "    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n",
    "  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n",
    "    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n",
    "\n",
    "from colabfold.colabfold import plot_protein\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# For some reason we need that to get pdbfixer to import\n",
    "if use_amber and '/usr/local/lib/python3.7/site-packages/' not in sys.path:\n",
    "    sys.path.insert(0, '/usr/local/lib/python3.7/site-packages/')\n",
    "\n",
    "def prediction_callback(unrelaxed_protein, length, prediction_result, input_features, type):\n",
    "  fig = plot_protein(unrelaxed_protein, Ls=length, dpi=150)\n",
    "  plt.show()\n",
    "  plt.close()\n",
    "\n",
    "result_dir=\".\"\n",
    "if 'logging_setup' not in globals():\n",
    "    setup_logging(Path(\".\").joinpath(\"log.txt\"))\n",
    "    logging_setup = True\n",
    "\n",
    "queries, is_complex = get_queries(queries_path)\n",
    "model_type = set_model_type(is_complex, model_type)\n",
    "download_alphafold_params(model_type, Path(\".\"))\n",
    "run(\n",
    "    queries=queries,\n",
    "    result_dir=result_dir,\n",
    "    use_templates=use_templates,\n",
    "    custom_template_path=custom_template_path,\n",
    "    use_amber=use_amber,\n",
    "    msa_mode=msa_mode,    \n",
    "    model_type=model_type,\n",
    "    num_models=5,\n",
    "    num_recycles=num_recycles,\n",
    "    model_order=[1, 2, 3, 4, 5],\n",
    "    is_complex=is_complex,\n",
    "    data_dir=Path(\".\"),\n",
    "    keep_existing_results=False,\n",
    "    recompile_padding=1.0,\n",
    "    rank_by=\"auto\",\n",
    "    pair_mode=pair_mode,\n",
    "    stop_at_score=float(100),\n",
    "    prediction_callback=prediction_callback,\n",
    "    dpi=dpi\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "\n",
    "(Double-click to edit the cell) <br>\n",
    "‚ùî __Question VIII: <u> What is pLDDT and how is the coloring and prediction related?</u>__  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question IX: <u> Why does AlphaFold2 needs to download the weights?</u>__  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question X: <u> Why does it run 5 models?</u>__  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XI: <u> What is different and what is simillar between the different predicted protein models?</u>__  \n",
    "_Answer:_  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualization of results\n",
    "The next cell creates an interactive 3D figure of the protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "KK7X9T44pWb7",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Display 3D structure \n",
    "import py3Dmol\n",
    "import glob\n",
    "import matplotlib.pyplot as plt\n",
    "from colabfold.colabfold import plot_plddt_legend\n",
    "\n",
    "jobname_prefix = \".custom\" if msa_mode == \"custom\" else \"\"\n",
    "if use_amber:\n",
    "  pdb_filename = f\"{jobname}{jobname_prefix}_relaxed_rank_{rank_num:03}_alphafold2_ptm_model_*.pdb\"\n",
    "else:\n",
    "  pdb_filename = f\"{jobname}{jobname_prefix}_unrelaxed_rank_{rank_num:03}_alphafold2_ptm_model_*.pdb\"\n",
    "\n",
    "pdb_file = glob.glob(pdb_filename)\n",
    "\n",
    "def show_pdb(rank_num=1, show_sidechains=False, show_mainchains=False, color=\"lDDT\"):\n",
    "  model_name = f\"rank_{rank_num}\"\n",
    "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',)\n",
    "  view.addModel(open(pdb_file[0],'r').read(),'pdb')\n",
    "\n",
    "  if color == \"lDDT\":\n",
    "    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n",
    "  elif color == \"rainbow\":\n",
    "    view.setStyle({'cartoon': {'color':'spectrum'}})\n",
    "  elif color == \"chain\":\n",
    "    chains = len(queries[0][1]) + 1 if is_complex else 1\n",
    "    for n,chain,color in zip(range(chains),list(\"ABCDEFGH\"),\n",
    "                     [\"lime\",\"cyan\",\"magenta\",\"yellow\",\"salmon\",\"white\",\"blue\",\"orange\"]):\n",
    "      view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n",
    "  if show_sidechains:\n",
    "    BB = ['C','O','N']\n",
    "    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
    "                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n",
    "                        {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
    "                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})  \n",
    "  if show_mainchains:\n",
    "    BB = ['C','O','N','CA']\n",
    "    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "\n",
    "  view.zoomTo()\n",
    "  return view\n",
    "\n",
    "\n",
    "show_pdb(rank_num,show_sidechains, show_mainchains, color).show()\n",
    "if color == \"lDDT\":\n",
    "  plot_plddt_legend().show() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "\n",
    "(Double-click to edit the cell)  \n",
    "‚ùî __Question XII: <u> What is the biological function of the intrinsically disordered region (IDR) of the protein, and is there a connection between the function and structure of this part of the protein? </u>__  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XIII: <u> What do you consider as the most intresting feature in the predicted model? </u>__  \n",
    "_Answer:_  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Prediction Quality\n",
    "To evaluate the prediction quality various prediction measurements can be visualised (e.g. PAE plots, sequence coverage, and predicted local distance difference test (plDDT)).  \n",
    "Run the cell below to visualize the quality of the predicted model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "11l8k--10q0C",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#@title Plots {run: \"auto\"}\n",
    "from IPython.display import display, HTML\n",
    "import base64\n",
    "from html import escape\n",
    "\n",
    "# see: https://stackoverflow.com/a/53688522\n",
    "def image_to_data_url(filename):\n",
    "  ext = filename.split('.')[-1]\n",
    "  prefix = f'data:image/{ext};base64,'\n",
    "  with open(filename, 'rb') as f:\n",
    "    img = f.read()\n",
    "  return prefix + base64.b64encode(img).decode('utf-8')\n",
    "\n",
    "pae = image_to_data_url(f\"{jobname}{jobname_prefix}_pae.png\")\n",
    "cov = image_to_data_url(f\"{jobname}{jobname_prefix}_coverage.png\")\n",
    "plddt = image_to_data_url(f\"{jobname}{jobname_prefix}_plddt.png\")\n",
    "display(HTML(f\"\"\"\n",
    "<style>\n",
    "  img {{\n",
    "    float:left;\n",
    "  }}\n",
    "  .full {{\n",
    "    max-width:100%;\n",
    "  }}\n",
    "  .half {{\n",
    "    max-width:50%;\n",
    "  }}\n",
    "  @media (max-width:640px) {{\n",
    "    .half {{\n",
    "      max-width:100%;\n",
    "    }}\n",
    "  }}\n",
    "</style>\n",
    "<div style=\"max-width:90%; padding:2em;\">\n",
    "  <h1>Plots for {escape(jobname)}</h1>\n",
    "  <h2> Predicted Aligned Error (PAE) Plots </h2>\n",
    "  <img src=\"{pae}\" class=\"full\" />\n",
    "  <img src=\"{cov}\" class=\"half\" />\n",
    "  <img src=\"{plddt}\" class=\"half\" />\n",
    "</div>\n",
    "\"\"\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview of output\n",
    "AlphaFold2 employs various metrics in the process of predicting protein structure models. Some of these metrics are depicted in the plots above and are briefly described below.\n",
    "\n",
    "### Predicted Aligned Error (PAE)  \n",
    "One of the outputs from AlphaFold2 is the Predicted Aligned Error (PAE) measured in √Öngstr√∂ms (capped at 31.75 √Ö). The PAE score for a given residue in the protein indicates the expected positional error between the predicted and experimental structures if the residues are aligned. The PAE scores represent the model confidence for the relative positions and orientations of different protein segments (e.g. two domains), for which reason rigid cores and protein domains possibly can be determined.  \n",
    "If the PAE scores are _low_ for a particular pair of residues, it means that the relative positions and orientations of the domains they belong to are predicted to be well-defined by AlphaFold. On the other hand, if the PAE values are _high_, it means that the relative position and orientation of the domains are predicted to be unreliable, and users should not consider any biological or structural relevance of these domains.\n",
    "\n",
    "### Sequence coverage\n",
    "Sequence coverage is a measure of how well a particular protein sequence has been studied and characterized. It is often depicted in the form of a plot, which shows the number of identified homologues (proteins that are similar in sequence to the protein of interest) along the length of the protein's representative sequence. The plot may also be colored to indicate the level of sequence identity between the protein of interest and each of the homologues. A high level of sequence coverage indicates that the protein has been well studied and characterized, while a low level of sequence coverage suggests that more research is needed to fully understand the protein's function and role in the cell. Sequence coverage can be useful for predicting the function of a protein, as it provides information about the protein's evolutionary history and the presence of conserved domains or motifs that may be important for its function.\n",
    "\n",
    "### Predicted Local Distance Difference Test (pLDDT)\n",
    "\n",
    "The predicted structure models produced by AlphaFold2 include atomic coordinates and a confidence estimate for each residue in the protein.The confidence measure is called the predicted Local Distance Difference Test (pLDDT) and is used to estimate the accuracy of the predicted CŒ± positions with experiemtanl structures. The pLDDT score is on a scale from 0 to 100, with higher scores indicating a higher confidence in the predicted structure. While pLDDT is a widely used metric in protein structure prediction, it is not clear how well it can be linked to physical properties such as protein dynamics. One of the main purposes of pLDDT is to assess the local accuracy of a prediction, which means that it can give high scores to well-predicted regions even if the overall prediction is not well-aligned with the true structure. This makes pLDDT a useful tool for identifying areas of the predicted structure that may be particularly reliable.\n",
    "\n",
    "\n",
    "üí°Useful link(s):  \n",
    "[AlphaFold Protein Structure Database: massively expanding the structural coverage of protein-sequence space with high-accuracy models\n",
    "](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728224/)  \n",
    "[Effective Molecular Dynamics from Neural-Network Based Structure Prediction Models](https://www.biorxiv.org/content/10.1101/2022.10.17.512476v1.full)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "\n",
    "(Double-click to edit the cell)  \n",
    "‚ùî __Question XIV: <u> In your own words, can you provide a explanation of what the Predicted Aligned Error (PAE) is and how it is visualized in a PAE plot?__</u>  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XV: <u> What are some interesting potential applications of AlphaFold2 that you can imagine?__</u>  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XVI: <u> Can you explain what information is depicted in the sequence coverage plot and how it relates to the N- and C-termini of the protein?__</u>   \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XVII: <u> Can you describe what the predicted Local Distance Difference Test (pLDDT) plot represents and how it relates to the predicted protein model?__</u>  \n",
    "_Answer:_  \n",
    "\n",
    "‚ùî __Question XVIII: <u> Elucidate the biological mechanisms underlying the relationship between the domains of the predicted protein structure and their functions, providing relevant examples.__</u>  \n",
    "_Answer:_  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "üí° Useful link(s):  \n",
    "[AlphaFold Error Estimates](https://www.rbvi.ucsf.edu/chimerax/data/pae-apr2022/pae.html)  \n",
    "[lDDT: a local superposition-free score for comparing protein structures and models using distance difference tests](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3799472/)  \n",
    "[Highly accurate protein structure prediction for the human proteome](https://www.nature.com/articles/s41586-021-03828-1)  \n",
    "[AlphaFold - FAQ](https://alphafold.ebi.ac.uk/faq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download the results\n",
    "You can download all your results by running the last cell below (not necessary)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "33g5IIegij5R",
    "tags": []
   },
   "outputs": [],
   "source": [
    "#@title Package and download results\n",
    "#@markdown If you are having issues downloading the result archive, try disabling your adblocker and run this cell again. If that fails click on the little folder icon to the left, navigate to file: `jobname.result.zip`, right-click and select \\\"Download\\\" (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n",
    "\n",
    "\n",
    "if msa_mode == \"custom\":\n",
    "  print(\"Don't forget to cite your custom MSA generation method.\")\n",
    "\n",
    "!zip -FSr $jobname\".result.zip\" config.json $jobname*\".json\" $jobname*\".a3m\" $jobname*\"relaxed_rank_\"*\".pdb\" \"cite.bibtex\" $jobname*\".png\"\n",
    "#files.download(f\"{jobname}.result.zip\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UGUBLzB3C6WN",
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Original Instructions from ColabFold (<font color='red'>not valid</font> for this version of ColabFold on UCloud)<a name=\"Instructions\"></a>\n",
    "**Quick start**\n",
    "1. Paste your protein sequence(s) in the input field.\n",
    "2. Press \"Runtime\" -> \"Run all\".\n",
    "3. The pipeline consists of 5 steps. The currently running step is indicated by a circle with a stop sign next to it.\n",
    "\n",
    "**Result zip file contents**\n",
    "\n",
    "1. PDB formatted structures sorted by avg. pLDDT and complexes are sorted by pTMscore. (unrelaxed and relaxed if `use_amber` is enabled).\n",
    "2. Plots of the model quality.\n",
    "3. Plots of the MSA coverage.\n",
    "4. Parameter log file.\n",
    "5. A3M formatted input MSA.\n",
    "6. A `predicted_aligned_error_v1.json` using [AlphaFold-DB's format](https://alphafold.ebi.ac.uk/faq#faq-7) and a `scores.json` for each model which contains an array (list of lists) for PAE, a list with the average pLDDT and the pTMscore.\n",
    "7. BibTeX file with citations for all used tools and databases.\n",
    "\n",
    "At the end of the job a download modal box will pop up with a `jobname.result.zip` file. Additionally, if the `save_to_google_drive` option was selected, the `jobname.result.zip` will be uploaded to your Google Drive.\n",
    "\n",
    "**MSA generation for complexes**\n",
    "\n",
    "For the complex prediction we use unpaired and paired MSAs. Unpaired MSA is generated the same way as for the protein structures prediction by searching the UniRef100 and environmental sequences three iterations each.\n",
    "\n",
    "The paired MSA is generated by searching the UniRef100 database and pairing the best hits sharing the same NCBI taxonomic identifier (=species or sub-species). We only pair sequences if all of the query sequences are present for the respective taxonomic identifier.\n",
    "\n",
    "**Using a custom MSA as input**\n",
    "\n",
    "To predict the structure with a custom MSA (A3M formatted): (1) Change the `msa_mode`: to \"custom\", (2) Wait for an upload box to appear at the end of the \"MSA options ...\" box. Upload your A3M. The first fasta entry of the A3M must be the query sequence without gaps. \n",
    "\n",
    "It is also possilbe to proide custom MSAs for complex predictions. Read more about the format [here](https://github.com/sokrypton/ColabFold/issues/76).\n",
    "\n",
    "As an alternative for MSA generation the [HHblits Toolkit server](https://toolkit.tuebingen.mpg.de/tools/hhblits) can be used. After submitting your query, click \"Query Template MSA\" -> \"Download Full A3M\". Download the A3M file and upload it in this notebook.\n",
    "\n",
    "**Using custom templates** <a name=\"custom_templates\"></a>\n",
    "\n",
    "To predict the structure with a custom template (PDB or mmCIF formatted): (1) change the `template_mode` to \"custom\" in the execute cell and (2) wait for an upload box to appear at the end of the \"Input Protein\" box. Select and upload your templates (multiple choices are possible).\n",
    "\n",
    "* Templates must follow the four letter PDB naming with lower case letters.\n",
    "\n",
    "* Templates in mmCIF format must contain `_entity_poly_seq`. An error is thrown if this field is not present. The field `_pdbx_audit_revision_history.revision_date` is automatically generated if it is not present.\n",
    "\n",
    "* Templates in PDB format are automatically converted to the mmCIF format. `_entity_poly_seq` and `_pdbx_audit_revision_history.revision_date` are automatically generated.\n",
    "\n",
    "If you encounter problems, please report them to this [issue](https://github.com/sokrypton/ColabFold/issues/177).\n",
    "\n",
    "**Comparison to the full AlphaFold2 and Alphafold2 colab**\n",
    "\n",
    "This notebook replaces the homology detection and MSA pairing of AlphaFold2 with MMseqs2. For a comparison against the [AlphaFold2 Colab](https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb) and the full [AlphaFold2](https://github.com/deepmind/alphafold) system read our [preprint](https://www.biorxiv.org/content/10.1101/2021.08.15.456425v1). \n",
    "\n",
    "**Troubleshooting**\n",
    "* Check that the runtime type is set to GPU at \"Runtime\" -> \"Change runtime type\".\n",
    "* Try to restart the session \"Runtime\" -> \"Factory reset runtime\".\n",
    "* Check your input sequence.\n",
    "\n",
    "**Known issues**\n",
    "* Google Colab assigns different types of GPUs with varying amount of memory. Some might not have enough memory to predict the structure for a long sequence.\n",
    "* Your browser can block the pop-up for downloading the result file. You can choose the `save_to_google_drive` option to upload to Google Drive instead or manually download the result file: Click on the little folder icon to the left, navigate to file: `jobname.result.zip`, right-click and select \\\"Download\\\" (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n",
    "\n",
    "**Limitations**\n",
    "* Computing resources: Our MMseqs2 API can handle ~20-50k requests per day.\n",
    "* MSAs: MMseqs2 is very precise and sensitive but might find less hits compared to HHblits/HMMer searched against BFD or MGnify.\n",
    "* We recommend to additionally use the full [AlphaFold2 pipeline](https://github.com/deepmind/alphafold).\n",
    "\n",
    "**Description of the plots**\n",
    "*   **Number of sequences per position** - We want to see at least 30 sequences per position, for best performance, ideally 100 sequences.\n",
    "*   **Predicted lDDT per position** - model confidence (out of 100) at each position. The higher the better.\n",
    "*   **Predicted Alignment Error** - For homooligomers, this could be a useful metric to assess how confident the model is about the interface. The lower the better.\n",
    "\n",
    "**Bugs**\n",
    "- If you encounter any bugs, please report the issue to https://github.com/sokrypton/ColabFold/issues\n",
    "\n",
    "**License**\n",
    "\n",
    "The source code of ColabFold is licensed under [MIT](https://raw.githubusercontent.com/sokrypton/ColabFold/main/LICENSE). Additionally, this notebook uses the AlphaFold2 source code and its parameters licensed under [Apache 2.0](https://raw.githubusercontent.com/deepmind/alphafold/main/LICENSE) and [CC BY 4.0](https://creativecommons.org/licenses/by-sa/4.0/) respectively. Read more about the AlphaFold license [here](https://github.com/deepmind/alphafold).\n",
    "\n",
    "**Acknowledgments**\n",
    "- We thank the AlphaFold team for developing an excellent model and open sourcing the software. \n",
    "\n",
    "- [KOBIC](https://kobic.re.kr) and [S√∂ding Lab](https://www.mpinat.mpg.de/soeding) for providing the computational resources for the MMseqs2 MSA server.\n",
    "\n",
    "- Richard Evans for helping to benchmark the ColabFold's Alphafold-multimer support.\n",
    "\n",
    "- [David Koes](https://github.com/dkoes) for his awesome [py3Dmol](https://3dmol.csb.pitt.edu/) plugin, without whom these notebooks would be quite boring!\n",
    "\n",
    "- Do-Yoon Kim for creating the ColabFold logo.\n",
    "\n",
    "- A colab by Sergey Ovchinnikov ([@sokrypton](https://twitter.com/sokrypton)), Milot Mirdita ([@milot_mirdita](https://twitter.com/milot_mirdita)) and Martin Steinegger ([@thesteinegger](https://twitter.com/thesteinegger)).\n"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "AlphaFold2.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
